client/src/main/java/org/apache/celeborn/client/read/LocalPartitionReader.java

```diff
@@ -0,0 +1,210 @@
+  private final AtomicReference<IOException> exception = new AtomicReference<>();
+  private final int fetchMaxReqsInFlight;
+  private final PartitionLocation location;
+  private volatile boolean closed = false;
+  private final int numChunks;
+  private int returnedChunks = 0;
+  private int currentChunkIndex = 0;
+  private final FileChannel shuffleChannel;
+  private List<Long> chunkOffsets;
+  private int endMapIndex;
+  private AtomicBoolean hasPendingFetchTask = new AtomicBoolean(false);
+
+  public LocalPartitionReader(
+      CelebornConf conf,
+      String shuffleKey,
+      PartitionLocation location,
+      TransportClientFactory clientFactory,
+      int startMapIndex,
+      int endMapIndex)
+      throws IOException {
+    fetchMaxReqsInFlight = conf.clientFetchMaxReqsInFlight();
+    results = new LinkedBlockingQueue<>();
+    this.location = location;
+    this.endMapIndex = endMapIndex;
+    PbStreamHandler streamHandle;
+    long fetchTimeoutMs = conf.clientFetchTimeoutMs();
+    try {
+      TransportClient client =
+          clientFactory.createClient(location.getHost(), location.getFetchPort(), 0);
+      TransportMessage openStreamMsg =
+          new TransportMessage(
+              MessageType.OPEN_STREAM,
+              PbOpenStream.newBuilder()
+                  .setShuffleKey(shuffleKey)
+                  .setFileName(location.getFileName())
+                  .setStartIndex(startMapIndex)
+                  .setEndIndex(endMapIndex)
+                  .setReadLocalShuffle(true)
+                  .build()
+                  .toByteArray());
+      ByteBuffer response = client.sendRpcSync(openStreamMsg.toByteBuffer(), fetchTimeoutMs);
+      streamHandle = TransportMessage.fromByteBuffer(response).getParsedPayload();
+    } catch (IOException | InterruptedException e) {
+      throw new IOException(
+          "Read shuffle file from local file failed, partition location: "
+              + location
+              + " filePath: "
+              + location.getStorageInfo().getFilePath(),
+          e);
+    }
+
+    chunkOffsets = new ArrayList<>(streamHandle.getChunkOffsetsList());
+    numChunks = streamHandle.getNumChunks();
+    shuffleChannel = FileChannelUtils.openReadableFileChannel(streamHandle.getFullPath());
+
+    logger.debug(
+        "Local partition reader {} offsets:{}",
+        location.getStorageInfo().getFilePath(),
+        StringUtils.join(chunkOffsets, ","));
+
+    if (hasPendingFetchTask.compareAndSet(false, true)) {
+      Future firstFetchTask = readLocalShufflePool.submit(() -> fetchDataTask());
+      try {
+        firstFetchTask.get();
+      } catch (Exception e) {
+        throw new CelebornIOException("Fetch data for local shuffle reader failed", e);
+      }
+    }
+    ShuffleClient.incrementLocalReadCounter();
+  }
+
+  private void fetchDataTask() {
+    try {
+      if (!closed && currentChunkIndex < numChunks && results.size() < fetchMaxReqsInFlight) {
```

Comment:
IMO we should avoid modifying `LocalPartitionReader`'s members in `fetchDataTask`, instead we should pass necessary information to `fetchDataTask`, for example
```
fetchDataTask(int startChunkIndex, int numChunks, ...)
```
and we should update `currentChunkIndex` after calling `submit`.